File: cs50x_credit/credit.c
#include <stdio.h>
#include <cs50.h>

bool luhn_function(long card_number);
string card_type(long card_number);

int main(void)
{
    long card_number = get_long("Number: ");

    if (luhn_function(card_number))
    {
        printf("%s\n", card_type(card_number));
    }
    else
    {
        printf("INVALID\n");
    }
}


bool luhn_function(long card_number)
{
    int sum = 0;
    bool alternate = false;

    while (card_number > 0)
    {
        int digit = card_number % 10;
        card_number /= 10;

        if (alternate)
        {
            digit *= 2;
            sum += (digit % 10) + (digit / 10);
        }
        else
        {
            sum += digit;
        }

        alternate = !alternate;
    }

    return (sum % 10) == 0;
}

string card_type(long card_number)
{
    int first_digit = 0, first_two_digits = 0, length = 0;
    long temp = card_number;

    while (temp > 0)
    {
        first_digit = temp % 10;
        if (temp < 100 && temp > 9)
        {
            first_two_digits = temp;
        }
        temp /= 10;
        length++;
    }

    if ((first_two_digits == 34 || first_two_digits == 37) && length == 15)
    {
        return "AMEX";
    }
    else if ((first_two_digits >= 51 && first_two_digits <= 55) && length == 16)
    {
        return "MASTERCARD";
    }
    else if (first_digit == 4 && (length == 13 || length == 16))
    {
        return "VISA";
    }
    return "INVALID";
}

File: cs50x_sentimental-credit/credit.py
from cs50 import get_int
import re

def main():
    card = get_int("Enter card number: ")
    card_str = str(card)
    check_sum = checksum(card_str)
    if check_sum:
        print(f"{checkdigits(card_str)}")
    else:
        print("INVALID")




#Check first two digits
def checkdigits(card_str):
    card_length = len(card_str)
    if re.match(r"^34|^37", card_str) and card_length == 15:
        return "AMEX"
    elif re.match(r"^51|^52|^53|^54|^55", card_str) and card_length == 16:
        return "MASTERCARD"
    elif re.match(r"^4", card_str) and (card_length == 13 or card_length == 16):
        return "VISA"
    else:
        return "INVALID"


#Perform Checksum

def checksum(card_str):
    total_sum = 0
    length = len(card_str)
    for i in range(length):
        digit = int(card_str[length - 1 - i])
        if i % 2 == 1:
            digit *= 2
            if digit > 9:
                digit -= 9
        total_sum += digit

    return total_sum % 10 == 0


main()
